File: packages/deep-engine/src/runtimePackage/dashboardTableTypes.ts
import type { DashboardRect } from "./dashboardCompositionTypes.js";

export interface DashboardFrozenTableV1 {
  readonly id: string;
  readonly pageId: string;
  readonly nodeIds: readonly string[];
  readonly title: string;
  /** One family without a select, otherwise one for each select option in the same order. */
  readonly families: readonly { readonly orders: readonly DashboardTableOrderV1[] }[];
}
export interface DashboardTableOrderV1 {
  readonly column: string | null;
  readonly direction: "asc" | "desc" | null;
  readonly exports: { readonly csv: string; readonly xlsx: string };
  readonly pages: readonly {
    readonly layers: readonly { readonly nodeId: string; readonly deep2d: string; readonly clip: DashboardRect | null;
      /** Placement of the slot content relative to the shared widget frame; keeps text geometry position-independent. */
      readonly origin?: readonly [number, number] }[];
    readonly controls: readonly {
      readonly action: "csv" | "xlsx" | "sort" | "previous" | "next";
      readonly column: string | null;
      readonly rect: DashboardRect;
      readonly enabled: boolean;
    }[];
  }[];
}

File: packages/deep-engine-native/src/runtime_package/dashboard_table_types.rs
use super::types::required_nullable;
use serde::{Deserialize, Serialize};
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct DashboardTable {
    pub id: String,
    pub page_id: String,
    pub node_ids: Vec<String>,
    pub title: String,
    pub families: Vec<DashboardTableFamily>,
}
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct DashboardTableFamily {
    pub orders: Vec<DashboardTableOrder>,
}
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct DashboardTableOrder {
    #[serde(deserialize_with = "required_nullable")]
    pub column: Option<String>,
    #[serde(deserialize_with = "required_nullable")]
    pub direction: Option<String>,
    pub exports: DashboardTableExports,
    pub pages: Vec<DashboardTableView>,
}
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct DashboardTableExports {
    pub csv: String,
    pub xlsx: String,
}
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct DashboardTableView {
    pub layers: Vec<DashboardTableLayer>,
    pub controls: Vec<DashboardTableControl>,
}
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct DashboardTableLayer {
    pub node_id: String,
    pub deep2d: String,
    #[serde(deserialize_with = "required_nullable")]
    pub clip: Option<[f64; 4]>,
    #[serde(default)]
    pub origin: Option<[f64; 2]>,
}
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct DashboardTableControl {
    pub action: String,
    #[serde(deserialize_with = "required_nullable")]
    pub column: Option<String>,
    pub rect: [f64; 4],
    pub enabled: bool,
}
